src/v/kafka/protocol/kafka_batch_adapter.cc

```diff
@@ -183,8 +187,105 @@ iobuf kafka_batch_adapter::adapt(iobuf&& kbatch) {
     return remainder;
 }
 
-void kafka_batch_adapter::adapt_with_version(iobuf kbatch, api_version) {
-    adapt(std::move(kbatch));
+/*
+ * Handle a MessageSet. For each uncompressed message the message data is
+ * accumulated in the new-style redpanda batch format. Compressed messages
+ * contain a nested MessageSet which is handled recursively (nesting should not
+ * exceed one level).
+ *
+ * The resulting record batch is compressed if any of the individual message
+ * sets are compressed.  In this case the compression type used is the last
+ * compression type encountered.
+ *
+ * The last message determines the timestamp used on the resulting record batch.
+ */
+void kafka_batch_adapter::convert_message_set(
+  storage::record_batch_builder& builder,
+  iobuf kbatch,
+  bool expect_uncompressed) {
+    auto parser = iobuf_parser(std::move(kbatch));
+    while (parser.bytes_left()) {
```

Comment:
I wonder if this should be futurized (probable a large knock-on effect, so another PR)

Reply:
I would have futurized this from the start, except the way this is integrated into the request flow is that its called deep inside a code generated parser that isn't futurized. If we hit a reactor stall I think we can build a workaround pretty quickly.

Reply:
good call. we should add a tix.

Reply:
https://github.com/vectorizedio/redpanda/issues/1396